Add introduction to the FAIR Data Pipeline (#6)
ScottishCovidResponse/hugo_source_files: content/docs/data_pipeline/introduction/_index.md

---
weight: 1
title: "Introduction"
---

# Introduction

**This is work in progress.**

> The FAIR Data Pipeline (FDP) is intended to enable tracking of provenance of data used in epidemiological modelling. File APIs written in C++, Java, Julia, Python and R can be called by model software for data insgestion. These interact with a local relational database storing metadata and the local filesystem, and are configured by a `yml` file associated with the model run. Local files and metadata are synchronised with a remote registry via a command line tool (`fdp`).

Reviewer: > Local files and metadata are synchronised with a remote registry via a command line tool (`fdp`).

Is this optional? Presumably there will be an option to run the registry locally without ever having to set up a remote registry, in the same way git does?

Reviewer 2: "can be"?

Reviewer 3: My only comment here is the "insgestion" typo.

> The FAIR Data Pipeline (FDP) is intended to enable tracking of provenance of data used in epidemiological modelling. File APIs written in C++, Java, Julia, Python and R can be called by model software for data insgestion. These interact with a local relational database storing metadata and the local filesystem, and are configured by a `yml` file associated with the model run. Local files and metadata are synchronised with a remote registry via a command line tool (`fdp`).

Reviewer: Just a small note, but for consistency we might want to decide on `yaml` vs `yml`

Reviewer 2: I personally prefer non-three-letter-abbreviations, I have to admit...

Reviewer 4: "configured by a yaml file" works best I's say. I'd only use 'yml' if I were directly referring to the file extension, '.yml'.

> The FAIR Data Pipeline (FDP) is intended to enable tracking of provenance of data used in epidemiological modelling. File APIs written in C++, Java, Julia, Python and R can be called by model software for data insgestion. These interact with a local relational database storing metadata and the local filesystem, and are configured by a `yml` file associated with the model run. Local files and metadata are synchronised with a remote registry via a command line tool (`fdp`).

Reviewer 2: As @soniamitchell says further down, we need a consistent name for the new API - you're using File API here, but that used to mean something different. I propose "Pipeline API" as a neutral term. Also typo:
```suggestion
The FAIR Data Pipeline (FDP) is intended to enable tracking of provenance of data used in epidemiological modelling. Pipeline APIs written in C++, Java, Julia, Python and R can be called by modelling software for data ingestion. These interact with a local relational database storing metadata and the local filesystem, and are configured using a yaml file associated with the model run. Local files and metadata can be synchronised with a remote registry via a command line tool (`fdp`).
```

> The FAIR Data Pipeline (FDP) is intended to enable tracking of provenance of data used in epidemiological modelling. File APIs written in C++, Java, Julia, Python and R can be called by model software for data insgestion. These interact with a local relational database storing metadata and the local filesystem, and are configured by a `yml` file associated with the model run. Local files and metadata are synchronised with a remote registry via a command line tool (`fdp`).

Reviewer 3: This seems fine on the C++ side, just the "insgestion" typo I noticed.

Author: Fixed, thanks!

The key benefits of using the FAIR Data Pipeline are:

> - Provenance of model outputs is tracked to inputs

Reviewer 2: ```suggestion
- Data recorded in a FAIR fashion (metadata on all data and code open and available for inspection)
- Provenance tracing allows model outputs to be traced to inputs and modelling code
```

Author: Thanks!

- Multiple language support

> - Designed to run on a broad range of platforms (including HPC)

Reviewer 2: ```suggestion
- Designed to run on a broad range of platforms (including HPC, inside Safe Havens)
- Designed to be set up and completed online (to down-/up-load data) and run offline (Safe Havens will require this)
```

Author: Thanks.

> - Access to shared central data for covid modelling

Reviewer: Again, perhaps not an important point, but do you mean to focus on COVID modelling? or epidemiological modelling? or perhaps statistical and mathematical modelling? 

I guess it depends on who this page is for and how closely you want to align with the grant outputs, whatever they are.

> - Access to shared central data for covid modelling

Reviewer 2: ```suggestion
- Open metadata provides knowledge of or access to shared central data for specific domains (e.g. COVID-19 epidemiological modelling)
```

Author: Thanks.

## Running Models

> To use the FAIR Data Pipeline with a piece of modelling software, you must add a language specific API as a dependency and interact with data registered in the pipeline via the methods it presents. Each model run must be configured using a [`config.yml`](../interface/_index.md) file which specifies inputs and outputs by metadata.

Reviewer: To be explicit and match the diagram below, do you want to say "you must add a language specific **File** API"? or perhaps "Local API"? (do you need both?) Is this the terminology we're using, not "Modelling API" or something else?

Reviewer 2: We originally used Standard API, and contrasted it with a File API as the low-level interface. I'd prefer to drop both now as too vague... I have been calling it the Modelling API, but that's wrong too (not always modelling). Maybe we call it the Pipeline API now? Then there is a separate Registry API, which is what the Pipeline API uses to interact with the registry.
```suggestion
To use the FAIR Data Pipeline with a piece of modelling software, you must add a language specific Pipeline API as a dependency and interact with data registered in the pipeline via the methods it presents. Each model run must be configured using a [`config.yml`](../interface/_index.md) file which specifies inputs and outputs by metadata.
```

Author: Thanks.

{{<mermaid align="left">}}

> graph LR;
>     subgraph Local
>         LR[Local Registry]
>         FS[File System]
>     end
>     subgraph Local API
>         API
>         CY
>     end
>     subgraph Model
>         M
>     end
> 
>     LR-->|read_*| API[File API]
>     API-->|write_*| LR
>     FS-->|read_*| API
>     API-->|write_*| FS
> 
>     CY[config.yml]-->API
> 
>     API-->M[Model]
>     M-->API

Reviewer 2: Slightly more complete?
```suggestion
graph LR;
    subgraph Local
        LR[Local Registry]
        FS[File System]
    end
    subgraph Local API
        API[Pipeline API]
        CY[config.yml]
    end
    subgraph Model
        M[Model]
    end
 
    LR -->|read_*| API
    API-->|write_*,link_*| LR
    FS-->|read_*| API
    API-->|write_*| FS

    CY-->API

    API-->|read_*,link_*|M
    M-->|write_*,link_*|API
    M-->|link_*|FS
```

Author: More complete but also more complicated-looking. May need to rethink later.

{{< /mermaid >}}

> 

Reviewer: This is a cool diagram but it's incomplete. Also, depending on who your target audience is you may wish to add some accompanying text.

Author: I believe @richardreeve 's proposed changes have fixed this for now.

## Getting data

The command line utility `fdp` is used to download and upload data and metadata required for and produced by model runs.

{{<mermaid align="left">}}
graph LR;
    subgraph Remote
        RR
        OS
        URI
    end
    subgraph Local
        LR
        FS
    end

    RR[Remote Registry]-->|fdp pull| LR[Local Registry]
    LR-->|fdp push| RR

    RR-->OS(Managed Object Store)
    RR-->URI(Abribtrary URI)
    LR-->FS(Local Filesystem)

    OS-->|fdp pull| FS
    URI-->|fdp pull| FS
    FS-->|fdp push| OS
    FS-->|fdp push?| URI

> {{< /mermaid >}}

Reviewer: Again, regardless of who the target audience is some explanation would be helpful here. 

Also, why "Arbitrary API"? and it might be clearer to use a consistent naming scheme for local and remote data stores.

Though I wonder if these diagrams might be more useful in a technical document, with a detailed breakdown of each function (such as [these](https://scottishcovidresponse.github.io/docs/data_pipeline/run_through/#fdp-pull))?

Reviewer 2: Arbitrary URI is fine I think - this is just data not stored in the default store - we haven't defined how that works yet though, so I think this will need more development over time.
